engines/router/missionctl/server/upi/server.go

```diff
@@ -0,0 +1,167 @@
+	}
+}
+
+func (us *Server) Run() {
+	s := grpc.NewServer()
+	upiv1.RegisterUniversalPredictionServiceServer(s, us)
+	reflection.Register(s)
+	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", us.port))
+	if err != nil {
+		log.Glob().Errorf("Failed to listen on port %d: %s", us.port, err)
+		return
+	}
+
+	errChan := make(chan error, 1)
+	stopChan := make(chan os.Signal, 1)
+
+	// bind OS events to the signal channel
+	signal.Notify(stopChan, syscall.SIGTERM, syscall.SIGINT)
+
+	go func() {
+		if err := s.Serve(listener); err != nil {
+			errChan <- err
+		}
+	}()
+
+	defer func() {
+		s.GracefulStop()
+	}()
+
+	// block until either OS signal, or server fatal error
+	select {
+	case err := <-errChan:
+		log.Glob().Errorf("Failed to start Turing Mission Control API: %s", err)
+	case <-stopChan:
+		log.Glob().Info("Signal to stop server")
+	}
+
+}
+
+func (us *Server) PredictValues(ctx context.Context, req *upiv1.PredictValuesRequest) (
+	*upiv1.PredictValuesResponse, error) {
+	var predictionErr *errors.TuringError // Measure execution time
+	defer metrics.Glob().MeasureDurationMs(
+		metrics.TuringComponentRequestDurationMs,
+		map[string]func() string{
+			"status": func() string {
+				return metrics.GetStatusString(predictionErr == nil)
+			},
+			"component": func() string {
+				return tracingComponentID
+			},
+		},
+	)()
+
+	// Create context from the request context
+	ctx = turingctx.NewTuringContext(ctx)
+	// Create context logger
+	ctxLogger := log.WithContext(ctx)
+	defer func() {
+		_ = ctxLogger.Sync()
+	}()
+
+	// if request comes with metadata, attach it to metadata to be sent with fiber
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		md = metadata.New(map[string]string{})
+	}
+
+	// Get the unique turing request id from the context
+	turingReqID, err := turingctx.GetRequestID(ctx)
+	if err != nil {
+		ctxLogger.Errorf("Could not retrieve Turing Request ID from context: %v",
+			err.Error())
+	}
+	ctxLogger.Debugf("Received request for %v", turingReqID)
+	md.Append(constant.TuringReqIDHeaderKey, turingReqID)
+
+	if tracing.Glob().IsEnabled() {
+		var sp opentracing.Span
+		sp, _ = tracing.Glob().StartSpanFromContext(ctx, tracingComponentID)
+		if sp != nil {
+			defer sp.Finish()
+		}
+	}
+
+	fiberRequest := &fibergrpc.Request{
+		Message:  req,
+		Metadata: md,
+	}
+	resp, predictionErr := us.getPrediction(ctx, fiberRequest)
+	if predictionErr != nil {
+		logTuringRouterRequestError(ctx, predictionErr)
+		return nil, predictionErr
+	}
+	return resp, nil
+}
+
+func (us *Server) getPrediction(
+	ctx context.Context,
+	fiberRequest fiber.Request) (
+	*upiv1.PredictValuesResponse, *errors.TuringError) {
+
+	// Create response channel to store the response from each step. 1 for route now,
+	// should be 4 when experiment engine, enricher and ensembler are added
+	respCh := make(chan grpcRouterResponse, 1)
+	defer close(respCh)
+
+	// Defer logging request summary
+	defer func() {
+		go logTuringRouterRequestSummary(
```

Comment:
i think there is a chance that respCh already closed and not all the value in the channel already logged